[Data] Add support for objects to Arrow blocks

terraflops1048576 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Currently, Ray does not support blocks/batches with objects and multi-dimensional arrays in different columns. This causes Ray Data to throw exceptions when these are provided because:
1. Since there's an arbitrary object in the batch, the Arrow block format fails with ArrowNotImplemented with dtype 17. This falls back to `return pd.DataFrame(dict(batch))` in `BlockAccessor.batch_to_block`.
2. However, this particular DataFrame constructor does not support columns with numpy.ndarray objects, so it throws the exception listed in the linked issue.

This change enables Python object storage in the Arrow blocks by defining an Arrow extension type that simply represents the Python objects as a variable-sized large binary. I suppose the alleged performance benefits listed in the comments are an extra benefit.

I'm not sure that this is the correct approach or that I've properly patched all of the places, so some help would be appreciated!

## Related issue number

Resolves #45235
## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

terraflops1048576 commented on python/ray/data/tests/test_arrow_serialization.py:
```diff
@@ -526,6 +541,22 @@ def fn(batch: list):
 
 
+def test_arrow_object_and_array_support(ray_start_regular_shared):
+    obj = ray.train.ScalingConfig(num_workers=4)
```
I just needed an object that wasn't subject to the vagaries of the importing from the test module, so I chose this.

bveeramani commented on python/ray/air/util/object_extensions/arrow.py:
```diff
@@ -0,0 +1,91 @@
+from ray._private.utils import _get_pyarrow_version
+from ray.util.annotations import PublicAPI
+
+MIN_PYARROW_VERSION_SCALAR_SUBCLASS = parse_version("9.0.0")
```
What happens if we use a lower version of PyArrow? Which APIs do we need?

terraflops1048576 replied:
Subclassing extension scalars is only available in PyArrow 9.0. This is needed for the code:
```
@PublicAPI(stability="alpha")
class ArrowPythonObjectScalar(pa.ExtensionScalar):
    def as_py(self) -> typing.Any:
        if not isinstance(self.value, pa.LargeBinaryScalar):
            raise RuntimeError(
                f"{type(self.value)} is not the expected LargeBinaryScalar"
            )
        return pickle.load(pa.BufferReader(self.value.as_buffer()))
```
that allows all of the as_py conversions in the array.

anyscalesam replied:
@bveeramani WDYT; OK (if so please resolve this conflict).

raulchen replied:
This should be fine.  We'll consider bumping the minimum PyArrow version later.

anyscalesam commented:
@raulchen is this ready to merge?

raulchen commented:
sorry, I got disruptions last week. doing the 2nd pass of review now. 
Also, could you resolve the conflicts when you have a chance? thanks

raulchen commented on python/ray/data/_internal/arrow_ops/transform_pyarrow.py:
```diff
@@ -74,30 +75,46 @@ def unify_schemas(
-                        f"but schema is not ArrayTensorType: {tensor_array_types[0]}"
-                    )
-                schema_field_overrides[col_name] = new_type
+    arrow_tensor_types = (ArrowVariableShapedTensorType, ArrowTensorType)
```
can you explain what this change is doing?

terraflops1048576 replied:
Basically this is a refactor of the unify_schemas function, which serves a similar purpose to the PyArrow unify_schemas. Its job is to take the schemas of the tables to be concatenated and produce a "unified" schema which contains the correct types for all of the columns of the constituent tables.

The previous code checked only the first table in the tables to concatenate, so it would behave weirdly if the first table weren't a tensor type and there was one further along by creating a schema where the tensor types were just erased (the error would be caught elsewhere, though).

However, for the object array, we need to support the case where the first table has a column of type, say, int64, and the second table has a column of type SomeObject. The old code would simply set the type to int64, which would cause an error. The refactored code will ensure that the int64s get pickled (yes, this is not very efficient, but this is the only way to support  So this new code scans all of the table schemas, finds which columns have any table that contain tensors or objects, and handles them accordingly.

This is basically type promotion for the ArrowTensorArray and the ArrowVariableShapedTensorArray and likewise for the ArrowPythonObjectArray and other types. Should we set the promotion options differently too?

raulchen replied:
Thanks for the detailed explanation! Can you also update the docstring of this method to reflect this change?
also do you mind explaining again what you mean by "set the promotion options differently"?

raulchen reviewed (commented):
Looks good to me at the high level.
Apologize for my late review. Also none of the active Ray Data maintainers are familiar with the pyarrow extension. it'd be appreciated if you can help explain the questions and add more comments. 
thanks.

terraflops1048576 commented on python/ray/air/util/object_extensions/arrow.py:
```diff
@@ -0,0 +1,91 @@
+
+def object_extension_type_allowed() -> bool:
+    return (
+        PYARROW_VERSION is None
```
Actually, I'm not sure. I just borrowed these version checks from the existing ArrowTensorArray implementation, which also passes if the PYARROW_VERSION is None. It turns out the way I've done it is buggy, so I'll revise it. I can also defensively just make this check fail if PYARROW_VERSION is None.

The more general question is whether we would be open to bumping the minimum PyArrow version to 9, which is almost 2 years old at this point. Currently, it is at 6.0.1, which appears to be almost 3 years old.

terraflops1048576 commented on python/ray/air/util/object_extensions/arrow.py:
```diff
@@ -0,0 +1,91 @@
+
+
+@PublicAPI(stability="alpha")
+class ArrowPythonObjectType(pa.ExtensionType):
```
Will do

terraflops1048576 commented on python/ray/data/_internal/arrow_block.py:
```diff
@@ -135,12 +138,29 @@ def __init__(self):
+            except (
+                pyarrow.ArrowInvalid,
+                pyarrow.ArrowNotImplementedError,
+                pyarrow.ArrowTypeError,
```
I believe this is the way that the fallback to Pandas is calculated; in any case, this code is outdated after the merge conflict because they have now all been coalesced into ArrowConversionError.

It is possible to check that all of the objects inside are of some Arrow type, but I believe this is quite cumbersome -- you have to manually check whether it is convertible to any PyArrow type, and you have to check all of the elements. It's easier to try conversion and see if it works.

In any case, the conversion into the ArrowPythonObjectArray should never fail if everything is pickleable, and I assume this is better than falling back to Pandas.

terraflops1048576 commented on python/ray/data/_internal/arrow_block.py:
```diff
@@ -207,7 +231,23 @@ def numpy_to_block(
+                        if log_once(f"arrow_object_pickle_{col_name}"):
+                            logger.warning(
+                                f"Failed to interpret {col_name} as "
+                                "multi-dimensional arrays. It will be pickled."
```
The current code looks at any column which has a numpy dtype of object and checks if it can be converted into a list of tensors. If it can't be, then we fall back to ArrowPythonObjectArray. This happens when the numpy array has a dtype of object because you have custom objects in there.

terraflops1048576 commented on python/ray/data/tests/test_arrow_block.py:
```diff
@@ -42,6 +43,39 @@ def test_register_arrow_types(tmp_path):
+    ds = ds.materialize()
+    block = ray.get(ds.get_internal_block_refs()[0])
+    # TODO: Once we support converting dict to a supported arrow type,
+    # the block type should be Arrow.
```
Ah yes, this used to be a test that checked the fallback behavior to Pandas blocks, but these changes are supposed to remove that fallback behavior for the case given. I cut the test from where it originally was and pasted it here, but I must've forgotten to delete the comment.

terraflops1048576 commented on python/ray/data/tests/test_transform_pyarrow.py:
```diff
@@ -181,6 +185,32 @@ def test_arrow_concat_tensor_extension_uniform_but_different():
     # fails for this case.
 
 
+@pytest.mark.skipif(
```
I'm not sure if these actually get run in CI if I put the skipif and require PyArrow version >= 9.0.0. Can someone check?

raulchen replied:
yes. it got run here https://buildkite.com/ray-project/microcheck/builds/2214

terraflops1048576 commented:
After trying to fix the test failures that relate to the new exception info, I realized that this is actually rather problematic -- it fixes way too many errors that would otherwise be reported to the user, like the overflow error resulting from [1, 2**100], trying to concatenate `np.array([2**100])` and `np.array([1])`, etc. After all, it is perfectly legal to pickle all of these things and store it in a column, but we obviously don't want to do this for performance reasons. We would rather error and tell the user to adjust the problematic values.

I'm not sure what we should do with the np.array case, as they are technically objects, but maybe we shouldn't fall back if anything related to ArrowTensorArray fails.

raulchen commented on python/ray/data/_internal/arrow_block.py:
```diff
@@ -52,6 +59,17 @@
+def is_object_fixable_error(e: ArrowConversionError) -> bool:
+    """Returns whether this error can be fixed by using an ArrowPythonObjectArray"""
+    return any(
+        err in "".join(traceback.format_exception(type(e), e, e.__traceback__))
```
nit, it's not efficient to format the entire error as a string. 
Can we iterate over the causes and check their types?

raulchen reviewed (commented):
Thanks for the new comments and explanation. 

> it fixes way too many errors that would otherwise be reported to the user, like the overflow error resulting from [1, 2**100], trying to concatenate np.array([2**100]) and np.array([1]), etc. After all, it is perfectly legal to pickle all of these things and store it in a column, but we obviously don't want to do this for performance reasons. 

To confirm my understanding, in the UDF return, if any value is overflowed, all values will fallback to objects. If nothing is overflowed, the type will still be integers. Right?

This seems fine. Because we don't know if the users intend to return overflowed integers either. Maybe we can introduce a flag to allow disabling the fallback.

> I'm not sure what we should do with the np.array case, as they are technically objects, but maybe we shouldn't fall back if anything related to ArrowTensorArray fails.

Are you concerned about np.arrays with object dtype? I think it's okay to not handle that. Users can just return a python list instead.

raulchen commented:
Some tests are failing. can you take a look? https://buildkite.com/ray-project/microcheck/builds/2214

terraflops1048576 commented:
I've tried to separate out overflows and some other errors like mismatched float/int types and make them continue reporting errors as in the tests I've fixed, instead of saving them as objects in the PyArrow table, primarily just to avoid unintended consequences.

You are correct that if we enabled catching all ArrowConversionErrors and turning things into objects that as long as blocks were completely non-overflowed integers, they would be stored as integer type. Currently, np.ndarrays with object dtype (such as when a user tries to put in overflowed integers) are just pickled as objects, because it's really hard to detect whether it's the result of overflowed integers or just plain objects in there.

raulchen commented on python/ray/data/_internal/arrow_block.py:
```diff
@@ -52,6 +58,17 @@
 
 T = TypeVar("T")
+logger = logging.getLogger(__name__)
+
+
+def is_object_fixable_error(e: ArrowConversionError) -> bool:
+    ARROW_OBJECT_FIXABLE_ERRORS = (
+        pyarrow.lib.ArrowTypeError,
+        pyarrow.lib.ArrowNotImplementedError,
+        pyarrow.lib.ArrowInvalid,
+    )
+    """Returns whether this error can be fixed by using an ArrowPythonObjectArray"""
+    return isinstance(e.__cause__, ARROW_OBJECT_FIXABLE_ERRORS)
```
```suggestion

ARROW_OBJECT_FIXABLE_ERRORS = (
    pyarrow.lib.ArrowTypeError,
    pyarrow.lib.ArrowNotImplementedError,
    pyarrow.lib.ArrowInvalid,
)


def is_object_fixable_error(e: ArrowConversionError) -> bool:
    """Returns whether this error can be fixed by using an ArrowPythonObjectArray"""
    return isinstance(e.__cause__, ARROW_OBJECT_FIXABLE_ERRORS)
```

raulchen reviewed (approved):
Thanks for the updates. LGTM

raulchen commented:
@terraflops1048576 the CI failures are unrelated and already fixed in master. Can you merge the latest master again?

terraflops1048576 commented:
I'm not sure why the tests that I fixed here weren't broken before, but I included a fix.

Interestingly enough, the [documentation](https://docs.ray.io/en/latest/data/api/doc/ray.data.Dataset.union.html) for `ray.data.Dataset.union` says

> The datasets must have the same schema as this dataset, otherwise the behavior is undefined.

However this isn't actually the case anymore and things like `python/ray/data/tests/test_consumption.py::test_union` actually depend on different behavior. Nevertheless, I fixed `unify_schemas` to reflect this behavior.
